hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/JobRepository.java

```diff
@@ -383,51 +399,102 @@ public JobResult getJobResult(long jobId) {
                   .sorted(comparing(JobResult::getCreationTime).reversed()).collect(toList());
     }
 
-    public static class FilterExecutionIdByJobIdPredicate implements Predicate<Long, Long>, IdentifiedDataSerializable {
+    /**
+     * Write the JobRecord to the IMap.
+     * <p>
+     * The write will be ignored if the timestamp of the given JobRecord is
+     * older than the timestamp of the stored record. See {@link
+     * UpdateJobRecordDynamicDataEntryProcessor#process}. It will also be ignored if the
+     * key doesn't exist in the IMap.
+     */
+    void writeJobRecordAsync(long jobId, DynamicData dynamicData) {
+        dynamicData.updateTimestamp();
+        jobRecords.submitToKey(jobId, new UpdateJobRecordDynamicDataEntryProcessor(jobId, dynamicData),
+                writeJobRecordCallback);
+    }
 
-        private long jobId;
+    /**
+     * See {@link #writeJobRecordAsync}.
+     */
+    void writeJobRecordSync(long jobId, DynamicData dynamicData) {
```

Comment:
Don't we have an isolation issue while using `writeJobRecordSync()` and `writeJobRecordAsync()` methods?

For instance, `JobRecord::ongoingSnapshotDone` method updates 7 fields of `JobRecord.DynamicData` and then calls `writeJobRecordSync()` to persist these changes.
In parallel, while `JobRecord::ongoingSnapshotDone` is updated some of the fields but not all of them, `writeJobRecordAsync()` method may kick in, say because of quorum value update, and persist a dirty `DynamicData` object.

Reply:
I don't get why we have the async version in the first place. It is used in `setJobStatusToString`, `finalizeJob`, and `updateQuorumSize` for rare cases and it think these places can use the sync version. It is also called in `beginSnapshot`, which seems to me as the most important place. Since it is called asyncly, some snapshot data can be persisted using the new `ongoingSnapshotId` value. However, the master node can fail before it persists the `DynamicData` object. If this is a safe state for us, then we don't need the async persistence call here. Then, we can get rid of the async job record persistence method completely.

Reply:
The synchronization is described in the second bullet in the [initial comment](#issue-221797573). The `MasterContext` has final `jobRecord` field. The `JobRecord` has volatile fields that are used to store the state and updates to them are properly synchronized. `MasterContext` instance is created on master as soon as the JobRecord is discovered through scanning or through a submit request from the client, and the same instance is kept around until the job completes (this includes while the job is suspended).

Thus, the `MasterContext.jobRecord` field is the _master replica_ of the record. We only read it initially and write it out only for the case should the master fail. That's why we can do it in an async way, and we do so by default. Since two async `put`s can be reordered, we use the `timestamp` to ensure that older async write doesn't overwrite a newer one already written.

We write the JobRecord in a sync way only to store the information that a snapshot is completed. The reason is that after this we're going to delete the previous successful snapshot and we don't want to start deleting before `JobRecord.dataMapIndex` is updated to refer to the map we're not going to delete.

If the master fails after it writes some data to ongoing snapshot, but before it increments `ongoingSnapshotId`, the new master always clears the inactive snapshot map when starting the job (MasterContext:319). There could be some zombie write from the old master using the same `snapshotId` as the new master is going to use for the next snapshot, but this should be detected by the snapshot validation.

Reply:
as I told above, async calls are already used in the rarely-called places, like switching to / returning from pending state. The only "critical" place async call made is `beginSnapshot` method. My understanding is, if a snapshot begins without ensuring the change in `JobRecord.DynamicData` is persisted and this is ok, we don't need this async call. Then, we can get rid of the async call in `beginSnapshot()`, and convert the others to sync calls. wdyt?

Reply:
I would much prefer sync calls and only way to update the record. `beginSnapshot()` shouldn't be critical either, it will only slightly delay starting of the snapshot.